### PR TITLE
Fixes #28521 - log location, pre/post, dynamic layout

### PR DESCRIPTION
**Pull request by lzap**
Cont. of https://github.com/theforeman/community-templates/pull/672

### CODE REVIEW

**Comment by ares** [comment]
Please re-render the snapshots.

File: app/views/unattended/partition_tables_templates/kickstart_dynamic.erb [hunk 1]
```diff
@@ -0,0 +1,47 @@
+<%#
+kind: ptable
+name: Kickstart dynamic
+model: Ptable
+oses:
+- CentOS
+- Fedora
+- RedHat
+%>
```
Comment (ekohl) [inline app/views/unattended/partition_tables_templates/kickstart_dynamic.erb]:
```suggestion
-%>
```

File: app/views/unattended/partition_tables_templates/kickstart_dynamic.erb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+<%#
+
+When a partition template contains line starting with "#Dynamic", the contents
+will be put into %pre shell scriplet executed with bash and it's expected it
+creates a file "/tmp/diskpart.cfg" which is then included into the Kickstart
+partitioning section. This allows generating parititioning scheme on the
```
Comment (ekohl) [inline app/views/unattended/partition_tables_templates/kickstart_dynamic.erb]:
```suggestion
partitioning section. This allows generating partitioning schemes on the
```

File: app/views/unattended/partition_tables_templates/kickstart_dynamic.erb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+oses:
+- CentOS
+- Fedora
+- RedHat
+%>
+<%#
+
+When a partition template contains line starting with "#Dynamic", the contents
+will be put into %pre shell scriplet executed with bash and it's expected it
+creates a file "/tmp/diskpart.cfg" which is then included into the Kickstart
```
Comment (ekohl) [inline app/views/unattended/partition_tables_templates/kickstart_dynamic.erb]:
```suggestion
will be put into %pre shell scriplet executed with bash and it's expected to
create a file "/tmp/diskpart.cfg" which is then included into the Kickstart
```

File: app/views/unattended/partition_tables_templates/kickstart_dynamic.erb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+name: Kickstart dynamic
+model: Ptable
+oses:
+- CentOS
+- Fedora
+- RedHat
+%>
+<%#
+
+When a partition template contains line starting with "#Dynamic", the contents
```
Comment (ekohl) [inline app/views/unattended/partition_tables_templates/kickstart_dynamic.erb]:
```suggestion
When a partition template contains a line starting with "#Dynamic", the contents
```

File: app/views/unattended/provisioning_templates/provision/atomic_kickstart_default.erb [hunk 5]
```diff
@@ -42,16 +42,51 @@ rootpw --iscrypted <%= root_pass %>
...
+%post --log=/root/install.post.custom.log
+# Custom post snippet generated by "<%= template_name + " custom post" %>"
+<%= snippet_if_exists(template_name + " custom post") %>
+<%= section_end -%>
+
+<%#
+The last post section halts Anaconda to prevent endless loop in case HTTP request fails
+%>
+<% if (is_fedora && os_major < 20) || (rhel_compatible && os_major < 7) -%>
+%post
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/provision/atomic_kickstart_default.erb:85]: It looks like you're using multiple `%post` sections. Is that allowed? If so, I learned something new today :)

File: app/views/unattended/provisioning_templates/provision/kickstart_default.erb [hunk 6]
```diff
@@ -319,8 +311,18 @@ sed -e 's/DEFAULTKERNEL=kernel-uek/DEFAULTKERNEL=kernel/g' -i /etc/sysconfig/ker
 touch /tmp/foreman_built
 <%= section_end -%>
 
+# copy %pre log files into chroot
+%post --nochroot
+cp -vf /tmp/*pre*log /tmp/sysimage/root/
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/provision/kickstart_default.erb]: Should it be stricter?
```suggestion
cp -vf /tmp/*.pre.*.log /tmp/sysimage/root/
```

File: app/views/unattended/provisioning_templates/provision/atomic_kickstart_default.erb [hunk 7]
```diff
@@ -42,16 +42,51 @@ rootpw --iscrypted <%= root_pass %>
 
 reboot
 
+<% if @dynamic -%>
+# Partition table dynamic generation
+%pre --log=/tmp/install.pre.dynamic.log
+<%= @host.diskLayout %>
+<%= section_end -%>
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/provision/atomic_kickstart_default.erb]: You need to set this variable. It's not in this file right now, unlike the default kickstart.

**Review by ekohl — COMMENTED** [review]
I wonder about the log locations. On the one hand, it makes debugging easier if it fails. On the other hand, it leaves more cruft in `/root`. We should document that in the release notes so users aren't surprised. This has some neat tricks that could be nice to show in a demo or a blog that the release notes refer to.

Some minor language things in comments inline, but overall the flow looks good.

File: app/views/unattended/provisioning_templates/provision/atomic_kickstart_default.erb (hunk 7 above)
Comment (lzap) [reply app/views/unattended/provisioning_templates/provision/atomic_kickstart_default.erb]: Nice find.

File: app/views/unattended/provisioning_templates/provision/atomic_kickstart_default.erb (hunk 5 above)
Comment (lzap) [reply app/views/unattended/provisioning_templates/provision/atomic_kickstart_default.erb:85]: Totally allowed and a good practice actually.

**Comment by lzap** [comment]
Rebased, fixed snapshot.

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Atomic Kickstart default.snap.txt [hunk 8]
```diff
@@ -18,6 +18,15 @@ rootpw --iscrypted xybxa6JUkz63w
...
+
+%pre --log=/tmp/install.pre.custom.log
+# Custom pre snippet generated by "Atomic Kickstart default custom pre"
+
+%end
+# copy %pre log files into chroot
+%post --nochroot
+cp -vf /tmp/*pre*log /tmp/sysimage/root/
+%end
 %post
```
Comment (ekohl) [inline test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Atomic Kickstart default.snap.txt]: The line endings between segments looks a bit odd. Maybe we want an empty line between various blocks? (repeated elsewhere, won't comment everywhere.

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart default.snap.txt [hunk 9]
```diff
@@ -41,22 +41,11 @@ wget
...
-/usr/bin/chvt 1
-) 2>&1 | tee /mnt/sysimage/root/install.postnochroot.log
+%post --nochroot --log=/mnt/sysimage/root/install.postnochroot.log
 %end
 
 %post --log=/root/install.post.log
 logger "Starting anaconda snapshothost postinstall"
-exec < /dev/tty3 > /dev/tty3
-#changing to VT 3 so that we can see whats going on....
-/usr/bin/chvt 3
```
Comment (ekohl) [inline test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart default.snap.txt:62]: I always found this to be useful to see what's going on. Especially on RHEL where pretty much everything is installed in `%post`, unlike CentOS where you actually have a progress indication. Please keep this in.

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart default.snap.txt [hunk 10]
```diff
@@ -41,22 +41,11 @@ wget
 redhat-lsb-core
 %end
 
-%post --nochroot
-exec < /dev/tty3 > /dev/tty3
-#changing to VT 3 so that we can see whats going on....
-/usr/bin/chvt 3
-(
-cp -va /etc/resolv.conf /mnt/sysimage/etc/resolv.conf
```
Comment (ekohl) [inline test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart default.snap.txt:53]: I kind of wonder why this was where. Looks like we now no longer do it, but I'm wondering why it was needed in the first place.

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart default.snap.txt (hunk 9 above)
Comment (lzap) [reply test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart default.snap.txt:62]: It does not work in EL8 where tmux is used.

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart default.snap.txt (hunk 10 above)
Comment (lzap) [reply test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart default.snap.txt:53]: No idea, it should not be there. We should stick with default Anaconda behavior.

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Atomic Kickstart default.snap.txt (hunk 8 above)
Comment (lzap) [reply test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Atomic Kickstart default.snap.txt]: Good find, we only have this in this file for some reason.

**Comment by lzap** [comment]
Rebased, if you insist on the tty3 I can probably do ERB IF-statement to keep it for EL7 and lower.

**Comment by ekohl** [comment]
Perhaps a quick poll on discourse would be useful to see if anyone cares?

**Comment by lzap** [comment]
> Perhaps a quick poll on discourse would be useful to see if anyone cares?

No need, if you like it that's enough to do it. Lemme rebase.

**Comment by lzap** [comment]
Correction, it does actually work even in RHEL8! I will fix that.

**Comment by lzap** [comment]
Rebased. Although it does not look like you actually can switch over to TTY3 and the console there is reserved for logs.

Screenshot_client-efi-nat_2020-10-13_14:00:50

File: app/views/unattended/provisioning_templates/provision/atomic_kickstart_default.erb [hunk 11]
```diff
@@ -3,6 +3,13 @@ kind: provision
 name: Atomic Kickstart default
 model: ProvisioningTemplate
 %>
+<%
+  rhel_compatible = @host.operatingsystem.family == 'Redhat' && @host.operatingsystem.name != 'Fedora'
+  is_fedora = @host.operatingsystem.name == 'Fedora'
+  os_major = @host.operatingsystem.major.to_i
+  os_minor = @host.operatingsystem.minor.to_i
+  section_end = (rhel_compatible && os_major <= 5) ? '' : '%end'
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/provision/atomic_kickstart_default.erb]: Is there even an Atomic based on EL5?

**Review by ekohl — DISMISSED** [review]

File: app/views/unattended/provisioning_templates/provision/atomic_kickstart_default.erb (hunk 11 above)
Comment (lzap) [reply app/views/unattended/provisioning_templates/provision/atomic_kickstart_default.erb]: Good point, removed for once and forever.

**Review by ekohl — DISMISSED** [review]
I think this looks good. Shall we merge it now so it's early in the release cycle?

**Comment by ekohl** [comment]
Also, because I took took a while there's a merge conflict now. Mind rebasing?

**Comment by lzap** [comment]
Sure, rebased.

**Review by ekohl — APPROVED** [review]
I moved the Redmine issue to the correct project so :+1: once Jenkins agrees.

**Comment by ekohl** [comment]
Katello looks unrelated to me.

**Pull request merged**